### PR TITLE
[docs] Reformat config reference

### PR DESCRIPTION
**Pull request by mattstein**
## The Problem/Issue/Bug:

The current [config options](https://ddev.readthedocs.io/en/stable/users/configuration/config_yaml/) page has several issues:

1. Its title is hard to parse, especially for a new user, and it doesn’t encapsulate the page contents.
2. It’s impossible to link directly to individual options, which is a critical problem for any reference material.
3. Options aren’t arranged in any order that helps the reader find a specific option without having to read every item on the page.
4. There’s no relationship between project and global options, so figuring out which have local/global equivalents is a matter of comparing tables.

## How this PR Solves The Problem:

1. **Re-titles the page** to “Config Options”, which should be more immediately obvious to understand, and places it at the top of its section navigation since available options are easily more important than the more specific subjects that follow.
2. **Introduces the concepts** of global and local configuration and how to work with them, along with reinforcing visual cues used throughout the page.
3. **Liberates options from their tables** into headings that can be linked directly.
4. **Alphabetizes** options which, combined with the sidebar table of contents, makes it possible to navigate the page without reading everything on it.
5. **Consolidates** options, consistently indicating whether they are per-project, global, or both. The option itself being more important than where you set it or how you use it, which should be a secondary concern.
6. **Links** descriptions to other settings or pages for supplemental learning and easier navigation in general.
7. **Clearly identifies** defaults and available options, where possible, for each option.

This format and ordering is more uniform and therefore easier to maintain with some confidence.

It also allows for a reasonable amount of explanation and linking for each item, which is discouraged by a table format where long descriptions hinder the ability to scan.

### What This Doesn’t Do

There’s always more that could be done, but I had to draw the line somewhere and leave things out:

- Include example settings (YAML and command, where relevant) for each item, which would reinforce how things work and make it hard to misunderstand.
- Further categorize options by area of concern, like formatting/output, project basics, debugging, etc.
- Improve option parity by consistently using `_` or `-` as word separators.
- Improve option parity by fixing inconsistent naming. (Booleans, for example, include `use_letsencrypt` vs. `mutagen_enabled` vs. `no_bind_mounts`. `use_`, `_enabled`, and `no_` are each different conventions.)
- Include a new, similarly-formatted and exhaustive reference page for commands.
- Rename `config_yaml.md` to `config-options.md`. (Hyphen separator and match title.)

#### Reduced Emphasis on YAML Location

This removes top-level project-vs.-global categorization on purpose. The availability of an option matters more initially than where you’d set it, and each config file itself includes exhaustive commentary on commonly-used settings.

The PR does make it harder, however, to limit your view to _only_ project or global options, which could ideally be solved with a custom template later on and some selector that could toggle between all (default), project, and global visibility.

#### Less Compact

The change in page format also makes it slightly less convenient to scan every local and global option name in one go. This is mitigated by the new use of the table of contents.

## Manual Testing Instructions:

https://ddev.readthedocs.io/en/stable/developers/testing-docs/#preview-changes

## Related Issue Link(s):

Once again related to #4188.

## Release/Deployment notes:

n/a

<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/4332"></a>

### CODE REVIEW

**Comment by mattstein** [comment]
@rfay Me again! You’ll definitely want to look at [the preview build](https://ddev--4332.org.readthedocs.build/en/4332/users/configuration/config_yaml/) for this one. PR description covers the rest.

**Comment by rfay** [comment]
Preview URL is https://ddev--4332.org.readthedocs.build/en/4332/ - no need to run mkdocs locally.

**Comment by mattstein** [comment]
@rfay That’s exactly what I linked to in the comment just above yours. I’ve linked to local testing instructions in every PR under “Manual Testing Instructions” for the sake of completeness.

File: docs/content/users/configuration/config_yaml.md [hunk 1]
```diff
@@ -1,76 +1,561 @@
...
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | `[]` |
+
+Example: `additional_fqdns: ["example.com", "sub1.example.com"]` would provide HTTP and HTTPS URLs for `example.com` and `sub1.example.com`.
+
+!!!warning
+    Take care with this; it adds items to your `/etc/hosts` file which can cause confusion.
+
+## `additional_hostnames`
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:50]: `ddev config --additional-hostnames=...` does work.

File: docs/content/users/configuration/config_yaml.md [hunk 2]
```diff
@@ -1,76 +1,561 @@
...
+
+You can override the per-project `config.yaml` with files named `config.*.yaml`.
+
+For example, many teams use `config.local.yaml` for configuration that’s specific to one environment, and not checked into the team’s default `config.yaml`.
+
+---
+
+## `additional_fqdns`
+
+An array of [extra fully-qualified domain names](../extend/additional-hostnames.md) to be used for a project.
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:39]: `ddev config --additional-fqdns` does work.

**Review by rfay — CHANGES_REQUESTED** [review]
This is absolutely over-the-pale wonderful.

It looks like there's a bit of work still to be done though, as almost all of the config options have `ddev config` equivalents, and they're not there yet.

**Comment by rfay** [comment]
Just can't say how fantastic this is.

**Comment by mattstein** [comment]
> It looks like there's a bit of work still to be done though, as almost all of the config options have ddev config equivalents, and they're not there yet.

That’s bullet one in my “What This Doesn’t Do” list above. There are a whole slew of things, some getting into DDEV’s naming in general, that’d make this a more useful reference and I was shooting for an MVP reformatting that page.

**Comment by mattstein** [comment]
I’m happy to add more to this effort, just want to know where to draw the line so it doesn’t get away from me or turn into a snowball PR.

**Comment by rfay** [comment]
I understand about the snowball... but it seems if you're going to do the tabs some places you'll want to do them everywhere?

**Comment by rfay** [comment]
Or maybe I should say you spoiled me in the first couple of paragraphs. I think readers will feel the same way.

**Comment by rfay** [comment]
Your call how to pursue it, but I think the fact that you spoiled me and set my expectations at the beginning will mean the same will happen to others.

**Comment by mattstein** [comment]
> Your call how to pursue it, but I think the fact that you spoiled me and set my expectations at the beginning will mean the same will happen to others.

I have other issues with tab inception and tab styling, and I’d love to come back to this to chase those remaining bullets later.

File: .spellcheckwordlist.txt [hunk 3]
```diff
@@ -434,6 +434,7 @@ not
 npm
 nssm
 nvm
+octicons
```
Comment (rfay) [inline .spellcheckwordlist.txt:437]: Is that a word?

**Comment by rfay** [comment]
OK, if you want to defer completion of the tabby stuff I'll read it without being disappointed.

**Comment by rfay** [comment]
For reasons that are beyond me, the config file items almost always have underscores and the command-line config stuff almost always has hyphens, so additional_fqdns and additional-fqdns. 

Is there any way in the world that we could ever make that seem rational or understood? I'm not sure that this is only in ddev, but it's not very helpful to the world.

File: .spellcheckwordlist.txt (hunk 3 above)
Comment (mattstein) [reply .spellcheckwordlist.txt:437]: Well, no. But the spell checker doesn’t like the `:octicons-*` convention and [it’s a real thing](https://squidfunk.github.io/mkdocs-material/reference/icons-emojis/?h=icons).

**Comment by mattstein** [comment]
> For reasons that are beyond me, the config file items almost always have underscores and the command-line config stuff almost always has hyphens, so additional_fqdns and additional-fqdns.
>
> Is there any way in the world that we could ever make that seem rational or understood? I'm not sure that this is only in ddev, but it's not very helpful to the world.

One way _around_ it is to explicitly show an example of each option—which is the right call even if it’s not a thing for this PR. This comes with the bonus benefit of omitting the explicit example for the few cases where a CLI argument isn’t an option—which also conveys something specific and important.

It also helps, I think, to go from “almost always have underscores” to “always”. Meaning YAML outliers like `table-style` should become `table_style`. That way, if you know the name of the setting and have a feel for YAML vs. CLI argument, you can pick the right name.

Same goes more subtly for settings that act like toggles. If we favored `_enabled`, for example:

- `mutagen_enabled` ⭐ 
- `instrumentation_opt_in` → `instrumentation_enabled`
- `no_bind_mounts` → `bind_mounts_enabled`
- `no_project_mount` → `project_mount_enabled`
- `use_hardened_images` → `hardened_images_enabled`
- `use_letsencrypt` → `letsencrypt_enabled`

There are probably other examples and `_enabled` may not be the right choice, but the idea’s the same: reward me for picking up on patterns by making things easier to understand and guess.

File: docs/content/users/configuration/config_yaml.md [hunk 4]
```diff
@@ -1,76 +1,561 @@
...
+
+When `true`, the `config.*.yaml` file with the option will have its settings *override* rather than *merge with* others. Allows statements like `nfs_mount_enabled: false` or `additional_hostnames: []` to work.
+
+## `php_version`
+
+The PHP version the project should use.
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | `8.0` | Can be `5.6`, `7.0`, `7.1`, `7.2`, `7.3`, `7.4`, `8.0`, `8.1`, or `8.2`.
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:340]: As these have been growing longer (it was once 5.6 and 7.0!) I've been simplifying with `5.6`-`8.2`

File: docs/content/users/configuration/config_yaml.md [hunk 5]
```diff
@@ -1,76 +1,561 @@
...
+
+Default TLD to be used for a project’s domains, or globally for all project domains.
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project<br>:octicons-globe-16: global | `ddev.site` | Can be changed to any TLD you’d prefer.
+
+## `required_docker_compose_version`
+
+Specific docker-compose version for download.
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:370]: This will only be used in specific troubleshooting cases. This option should not be used.

File: docs/content/users/configuration/config_yaml.md [hunk 6]
```diff
@@ -1,76 +1,561 @@
...
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-globe-16: global | `` |
+
+If set to `v2.8.0`, for example, it will download and use that version instead of the expected version for docker-compose.
+
+## `router_bind_all_interfaces`
+
+Whether to bind on all network interfaces.
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:380]:
```suggestion
Whether to bind `ddev-router`'s ports on all network interfaces.
```

File: docs/content/users/configuration/config_yaml.md [hunk 7]
```diff
@@ -1,76 +1,561 @@
...
+| -- | -- | --
+| :octicons-file-directory-16: project | `true` | Can be `true` or `false`.
+
+When `false`, DDEV will always update the `/etc/hosts` file with the project hostname instead of using DNS for name resolution.
+
+See [Using DDEV Offline](../details/offline-usage.md).
+
+## `use_docker_compose_from_path`
+
+Whether to use the system-installed docker-compose. You can otherwise use [`required_docker_compose_version`](#required_docker_compose_version) to specify a version for download.
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:470]: This should only be used in specific troubleshooting scenarios.

File: docs/content/users/configuration/config_yaml.md [hunk 8]
```diff
@@ -1,76 +1,561 @@
...
+
+## `override_config`
+
+Whether to override config values instead of merging.
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | `false` | Can be `true` or `false`.
+
+When `true`, the `config.*.yaml` file with the option will have its settings *override* rather than *merge with* others. Allows statements like `nfs_mount_enabled: false` or `additional_hostnames: []` to work.
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:332]: Please link to https://ddev.readthedocs.io/en/latest/users/extend/customization-extendibility/#extending-configyaml-with-custom-configyaml-files

File: docs/content/users/configuration/config_yaml.md [hunk 9]
```diff
@@ -1,76 +1,561 @@
...
+
+## `upload_dir`
+
+Path from the project’s docroot to the user-generated files directory targeted by `ddev import-files`.
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | |
+
+This is a CMS-specific setting that won’t have any effect for generic `php` [project types](#type).
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:454]:
```suggestion
```

That was obsolete.

File: docs/content/users/configuration/config_yaml.md [hunk 10]
```diff
@@ -1,76 +1,561 @@
...
+
+Example: `working_dir: { web: "/var/www", db: "/etc" }` sets the working directories for the `web` and `db` containers.
+
+## `xdebug_enabled`
+
+Whether Xdebug should be enabled for [step debugging](../debugging-profiling/step-debugging.md) or [profiling](../debugging-profiling/xdebug-profiling.md).
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | `false` | Can be `true` or `false`. Most people use `ddev xdebug` and `ddev xdebug off` commands instead of configuring this, because of Xdebug’s significant performance impact.
```
Comment (rfay) [inline docs/content/users/configuration/config_yaml.md:551]:
```suggestion
| :octicons-file-directory-16: project | `false` | Please leave this `false` in most cases. Most people use `ddev xdebug` and `ddev xdebug off` commands.
```

**Review by rfay — CHANGES_REQUESTED** [review]
It's just awesome, thanks!

File: docs/content/users/configuration/config_yaml.md [hunk 11]
```diff
@@ -1,76 +1,561 @@
...
+!!!note
+    DDEV v1.18 and earlier used `mariadb_version` and `mysql_version`, which are now automatically converted to the `database` format.
+
+## `dbimage_extra_packages`
+
+Extra Debian packages for the project’s database container.
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | `[]` |
```
[the remark this answers is not in the document]
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:95]: > This is great communication in rendered page (the images). I wish it could be meaningful in the source though.

Same. In VuePress world we could introduce custom elements like `<icon kind="globe"/>` or whatever. Not sure if there’s a better way here.

> I'm thinking that "empty" might be more descriptive than []...

I thought `[]` indicated an empty _array_ whereas the word “empty” is less specific about the expected format. But this isn’t a hill I’d like to die on.

File: docs/content/users/configuration/config_yaml.md [hunk 12]
```diff
@@ -1,76 +1,561 @@
...
+
+DDEV-specific lifecycle [hooks](hooks.md) to be executed.
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | `` |
+
+## `host_db_port`
+
+Port for binding database server to localhost interface.
```
[the remark this answers is not in the document]
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:161]: IIRC there are a few places where “automatic” should either be listed in default or clarified in the description. 🤔

File: docs/content/users/configuration/config_yaml.md (hunk 4 above)
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:340]: You okay leaving these spelled out? I don’t think it hurts the page layout, and the full list is both explicit and easy to copy from. Could save two people from trying PHP 7.6. 😄

File: docs/content/users/configuration/config_yaml.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:161]: [Fixed](https://github.com/drud/ddev/pull/4332/commits/7843b63505b6626984b31e8cb8745e8b08a2ec84).

File: docs/content/users/configuration/config_yaml.md [hunk 13]
```diff
@@ -1,76 +1,561 @@
...
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | |
+
+Can be a specific port number for a fixed database port, which can be useful for configuration of host-side database clients. (May still be easier to use `ddev mysql`, `ddev psql`, `ddev sequelpro`, etc., which handle changing ports automatically, as does the sample command `ddev mysqlworkbench`.)
+
+## `host_https_port`
+
+Specific, persistent HTTPS port for direct binding to localhost interface.
```
[the remark this answers is not in the document]
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:173]: [Fixed](https://github.com/drud/ddev/pull/4332/commits/7843b63505b6626984b31e8cb8745e8b08a2ec84).

File: docs/content/users/configuration/config_yaml.md [hunk 14]
```diff
@@ -1,76 +1,561 @@
...
+| -- | -- | --
+| :octicons-file-directory-16: project | |
+
+Not commonly used. Can be a specific port number for a fixed HTTPS URL. If unset, the port will change each time `ddev start` is run.
+
+Example: `59001` will have the project always use `https://127.0.0.1:59001` for the localhost URL—used less commonly than the named URL which is better to rely on.
+
+## `host_webserver_port`
+
+Specific, persistent HTTP port for direct binding to localhost interface.
```
[the remark this answers is not in the document]
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:185]: [Fixed](https://github.com/drud/ddev/pull/4332/commits/7843b63505b6626984b31e8cb8745e8b08a2ec84).

File: docs/content/users/configuration/config_yaml.md (hunk 4 above)
Comment (rfay) [reply docs/content/users/configuration/config_yaml.md:342]: yes, I'm ok with it.

File: docs/content/users/configuration/config_yaml.md (hunk 5 above)
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:372]: [Fixed](https://github.com/drud/ddev/pull/4332/commits/5258a6541398d0521dde8f64b5b151bacbf53191).

File: docs/content/users/configuration/config_yaml.md (hunk 7 above)
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:474]: [Fixed](https://github.com/drud/ddev/pull/4332/commits/5258a6541398d0521dde8f64b5b151bacbf53191).

File: docs/content/users/configuration/config_yaml.md (hunk 8 above)
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md:334]: [Done](https://github.com/drud/ddev/pull/4332/commits/2b85a55ad64b9339f194cbd6a9e1bcb9d80a3634).

File: docs/content/users/configuration/config_yaml.md [hunk 15]
```diff
@@ -1,76 +1,561 @@
...
+| -- | -- | --
+| :octicons-file-directory-16: project | `2` | Can be `2`, `1`, or empty (`""`) for latest major version at container build time.<br><br>Can also be a minor version like `2.2` for the latest release of that branch, an explicit version like `1.0.22`, or a keyword like `stable`, `preview` or `snapshot`. See Composer documentation.
+
+## `database`
+
+The type and version of the database engine the project should use.
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project | MariaDB 10.3 | Can be MariaDB 5.5–10.7, MySQL 5.5–8.0, or PostgreSQL 9–14.<br>See [Database Server Types](../extend/database_types.md) for examples and caveats.
```
[the remark this answers is not in the document]
Comment (mattstein) [reply docs/content/users/configuration/config_yaml.md]: I think [I figured it out](https://github.com/drud/ddev/pull/4332/commits/07d179174bc816ab2d96c34edd089b83a4e7ba1b)!

**Comment by rfay** [comment]
Let me know when this is how you want it. I'll take one more pass and pull.

**Comment by rfay** [comment]
Do it the way you think is best and we'll see if people stumble on it.

**Comment by mattstein** [comment]
Okay, ready to go then!
